Quick Start Guide (#502)
kokkos/kokkos-core-wiki: docs/source/quick_start.rst

Quick Start
============

This guide is intended to jump start new Kokkos users (and beginners, in particular).

Download Latest and Build 
-----------------------------

.. note::

  Please become familiar with `Kokkos Requirements <https://kokkos.org/kokkos-core-wiki/requirements.html>`_, and verify that your machine has all necessary compilers, backend GPU SDK (e.g., CUDA, ROCM, Intel oneAPI, etc.),and build system components.

..
 Nota bene:  the link for "Latest" should be stable from one release to the next, but check periodically to be sure 
..

:bdg-link-primary:`Latest Release <https://github.com/kokkos/kokkos/releases/latest>`

.. code-block:: sh
  
  # Uncomment according to the type of file you've downloaded (zip or tar)
  unzip kokkos-x.y.z.zip 
  # tar -xzf kokkos-x.y.z.tar.gz
  cd kokkos-x.y.z

Basic Configuration
-------------------

.. note::

>   You can create small shell scripts to manage and experiment with configuration details, following the GPU microarchitecture-appropriate examples below.  Upon successful configuration, ``make `` to build, and ``make install`` to install.

Reviewer: This doesn't render correctly.

OpenMP (CPU Parallelism)
~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: sh

>   cmake -B <name-of-build-directory> -DKokkos_ENABLE_OPENMP=ON -DCMAKE_BUILD_TYPE=Release -S <name-of-source-directory>

Reviewer: All of these look too long with respect to the line length.

.. note::

>   Kokkos will attempt to autodetect GPU microarchitecture, but it is also possible to specify the desired `GPU architecture <https://kokkos.org/kokkos-core-wiki/keywords.html#gpu-architectures>`_.  In scenarios where a backend (e.g., CUDA, HIP) is enabled, Kokkos will execute serially on the "host" (CPU).  

Reviewer 2: There is a slight bug in the wording here, it implies that if a backend is enabled Kokkos will run serially. We really mean that if a device backend is enabled, the default on the host will be serial

```suggestion
  Kokkos will attempt to autodetect GPU microarchitecture, but it is also possible to specify the desired `GPU architecture <https://kokkos.org/kokkos-core-wiki/keywords.html#gpu-architectures>`_.  In scenarios where a device (GPU) backend (e.g., CUDA, HIP) is enabled, Kokkos will default to serial execution on the host (CPU).
```

Reviewer 2: you can keep the quotes if you prefer but if you add them back add to "device" also

CUDA
~~~~

.. code-block:: sh

  cmake -B <name-of-build-directory> -DKokkos_ENABLE_CUDA=ON -DCMAKE_BUILD_TYPE=Release -S <name-of-source-directory>

HIP
~~~

.. code-block:: sh

  cmake -B <name-of-build-directory> -DKokkos_ENABLE_HIP=ON -DCMAKE_CXX_COMPILER=hipcc -DCMAKE_BUILD_TYPE=Release -S <name-of-source-directory>

Building and Linking a Kokkos "Hello World"
-------------------------------------------

.. note::

>   ``Kokkos_ROOT`` and the root directory for you target backend SDK (i.e., ``CUDA_ROOT``, ``ROCM_PATH``) will need to be set.  ``Kokkos_ROOT`` should be set to the path of your Kokkos installation.  If a modules environment, the SDK variables will be automatically set upon module loading (e.g., ``module load rocm/5.7.1).  Please see `Build, Install and Use <https://kokkos.org/kokkos-core-wiki/building.html>`_ for additional details.  The example detailed below is in the Kokkos Core `example` directory.

Reviewer 2: ```suggestion
  ``Kokkos_ROOT`` and the root directory for you target backend SDK (i.e., ``CUDA_ROOT``, ``ROCM_PATH``) will need to be set.  ``Kokkos_ROOT`` should be set to the path of your Kokkos installation.  In a modules environment, the SDK variables will be typically automatically set upon module loading (e.g., ``module load rocm/5.7.1).  Please see `Build, Install and Use <https://kokkos.org/kokkos-core-wiki/building.html>`_ for additional details.  The example detailed below is in the Kokkos Core `example` directory.
```

.. code-block:: sh

  git clone https://github.com/kokkos/kokkos.git 
  cd example/build_cmake_installed
  cmake -B <build directory> -S . -DKokkos_ROOT=<path-to-installation>
  cd <build directory>
  make 
  ./example
  

Getting Help
------------

If you need addtional help getting started, please join the `Kokkos Slack Channel <https://kokkosteam.slack.com>`_.  Here are `sign up details <https://kokkos.org/kokkos-core-wiki/faq.html#faq>`_.  Joining Kokkos Slack is the on ramp for becoming a project contributor.

> ..
>   **THIS IS A COMMENT**
>   *TODO*
>      - Integrate (merged) Quick Start with Cédric's PR:  https://github.com/kokkos/kokkos/pull/6796
>      - Ongoing reconciling with the Julien B. / KUG23- initiated discussion:  https://github.com/kokkos/internal-documents/pull/19
>      - Add `git submodule` "how to" for Kokkos
>      - Add Quick Start to main Kokkos page, such that it is the first thing you encounter on the landing page (kokkos.org)
>      - In V2, put the recipes for the different backends on different pages
>      - Julien B. suggested using github templates for the View "Hello World" example
>      - Add ``kokkos/example`` Nic M.:  CUDA as a CMake language example (using View): cmake -S . -B build -DKokkos_ENABLE_CUDA=ON CMAKE_CUDA_COMPILER=nvcc Kokkos_ENABLE_COMPILE_AS_CMAKE_LANGUAGE=ON [-DCMAKE_BUILD_TYPE=Release]
> ..

Reviewer 3: ```suggestion
```

Quick Start
============

This guide is intended to jump start new Kokkos users (and beginners, in particular).

Download Latest and Build 
-----------------------------

.. note::

  Please become familiar with `Kokkos Requirements <https://kokkos.org/kokkos-core-wiki/requirements.html>`_, and verify that your machine has all necessary compilers, backend SDK and build system components.

..
 Nota bene:  the link for "Latest" should be stable from one release to the next, but check periodically to be sure 

> ..

Reviewer 4: What's the note for?

Quick Start
============

> This guide is intended to jump start new Kokkos users (and beginners, in particular).  Kokkos Core is the foundation for all other Kokkos Ecosystem projects, and is readily usable in your own for performance portability.

Reviewer 4: Not sure what the 2nd sentence bring here.

Download Latest and Build 
-----------------------------

.. note::

>   Please become familiar with `Kokkos Requirements <https://kokkos.org/kokkos-core-wiki/requirements.html>`_, and verify that your machine has all necessary compilers, backend SDK and build system components.

Reviewer 4: What about 

Please ensure that your system meets the `minimum requirements <https://kokkos.org/kokkos-core-wiki/requirements.html>`_ (cmake and compiler toolchain)

Reviewer 4: Ping.  "[V]erify that your machine has all necessary compilers, backend SDK and build system components" is somewhat cryptic.

..
 Nota bene:  the link for "Latest" should be stable from one release to the next, but check periodically to be sure 
..

:bdg-link-primary:`Latest Release <https://github.com/kokkos/kokkos/releases/latest>`

.. code-block:: sh
  
  # Uncomment according to the type of file you've downloaded (zip or tar)
  unzip kokkos-x.y.z.zip 
  # tar -xzf kokkos-x.y.z.tar.gz
  cd kokkos-x.y.z
  mkdir build
  cd build

Basic Configuration
-------------------

.. note::

>   You can create small shell scripts to manage and experiment with configuration details, following the GPU microarchitecture-appropriate examples below.  Upon successful configuration, ``make -j`` to build, and ``make install`` to install.

Reviewer 4: Prefer `cmake --build BUILD-DIR` see my previous comments

Reviewer 4: Ping.  We should not suggest to use `make -j`.  As Cedric suggested this is less portable because it assumes that the Unix Makefiles generator is being used.  Also the `-j` option can be problematic (on a login node for instance) because it will use as many processes/jobs as it can detect from where the command is being run.

OpenMP (CPU Parallelism)
~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: sh

  cmake \
   -DKokkos_ENABLE_OPENMP=ON \
   -DCMAKE_BUILD_TYPE=Release ..

.. note::

  Kokkos will attempt to autodetect GPU microarchitecture, but it is also possible to specify the desired `GPU architecture <https://kokkos.org/kokkos-core-wiki/keywords.html#gpu-architectures>`_.  In scenarios where a backend (e.g., CUDA, HIP) is enabled, Kokkos will execute serially on the "host" (CPU).  

CUDA (CPU and GPU)
~~~~~~~~~~~~~~~~~~

.. code-block:: sh

  cmake \
    -DKokkos_ENABLE_CUDA=ON \
    -DCMAKE_BUILD_TYPE=Release ..

HIP (CPU and GPU)
~~~~~~~~~~~~~~~~~

.. code-block:: sh

  cmake \
    -DKokkos_ENABLE_HIP=ON \
    -DCMAKE_CXX_COMPILER=hipcc \
    -DCMAKE_BUILD_TYPE=Release ..

Basic Configuration Examples for Spack Builds

> ----------------------------------------------

Reviewer 4: It might be obvious to you but some might not realize that this is meant as an alternative to the previous section.
Also you could just refer to the spack section in the build reference as an alternative solution to install Kokkos and leave it there

Quick Start
============

This guide is intended to jump start new Kokkos users (and beginners, in particular).  Kokkos Core is the foundation for all other Kokkos Ecosystem projects, and is readily usable in your own for performance portability.

Download Latest and Build 
-----------------------------

.. note::

  Please become familiar with `Kokkos Requirements <https://kokkos.org/kokkos-core-wiki/requirements.html>`_, and verify that your machine has all necessary compilers, backend SDK and build system components.

..
 Nota bene:  the link for "Latest" should be stable from one release to the next, but check periodically to be sure 
..

:bdg-link-primary:`Latest Release <https://github.com/kokkos/kokkos/releases/latest>`

.. code-block:: sh
  

>   # Uncomment according to the type of file you've downloaded (zip or tar)
>   unzip kokkos-x.y.z.zip 
>   # tar -xzf kokkos-x.y.z.tar.gz

Reviewer 5: ```suggestion
  tar xzf kokkos-x.y.z.tar.gz
```

I think we can focus on one type of file, not having to add 2 other lines that make thing looks complicated.

Author: I put zip and tar.gz, because those are the choices you are presented on the latest release.  Please keep in mind that the "Quick Start" needs to be geared for the near-absolute beginner.

  cd kokkos-x.y.z

>   mkdir build
>   cd build

Reviewer 5: ```suggestion
```

Author: Again, I want to show a step-by-step recipe, leaving nothing to chance.  We are writing the "Quick Start" for people who have little experience.

Reviewer 4: See https://github.com/kokkos/kokkos-core-wiki/pull/502/files#r1495067795

Basic Configuration
-------------------

.. note::

>   You can create small shell scripts to manage and experiment with configuration details, following the GPU microarchitecture-appropriate examples below.  Upon successful configuration, ``make -j`` to build, and ``make install`` to install.

Reviewer 5: ```suggestion
  You can create small shell scripts to manage and experiment with configuration details, following the GPU microarchitecture-appropriate examples below. 
```
It is better to use `cmake` to run compile and install. Default generator might not be `make` but `ninja` or other stuff on windows.

Author: Following the same logic, that the "Quick Start" is for new users and/or people with little experience, I want to clearly separate configure and build steps, and on Linux-like systems, I *believe* one would mostly use `make`.  If you wanted to provide recipes for Windows / MSVC, I'd be happy to include those, I'm just not well versed in Windows as a software development environment.

   -DCMAKE_BUILD_TYPE=Release ..

.. note::

  Kokkos will attempt to autodetect GPU microarchitecture, but it is also possible to specify the desired `GPU architecture <https://kokkos.org/kokkos-core-wiki/keywords.html#gpu-architectures>`_.  In scenarios where a backend (e.g., CUDA, HIP) is enabled, Kokkos will execute serially on the "host" (CPU).  

CUDA (CPU and GPU)
~~~~~~~~~~~~~~~~~~

.. code-block:: sh

  cmake \
    -DKokkos_ENABLE_CUDA=ON \
    -DCMAKE_BUILD_TYPE=Release ..

HIP (CPU and GPU)
~~~~~~~~~~~~~~~~~

.. code-block:: sh

  cmake \
    -DKokkos_ENABLE_HIP=ON \
    -DCMAKE_CXX_COMPILER=hipcc \
    -DCMAKE_BUILD_TYPE=Release ..

Basic Configuration Examples for Spack Builds
----------------------------------------------

|br|
Diverse Kokkos variants can be built / installed via Spack.  You will need to select a variant for the desired backend, and appropriate GPU microarchitecture.  To explore the range of variants for a package, ``spack info kokkos``, ``spack info trilinos``, etc.  Please see basic `Spack installation  <https://spack.readthedocs.io/en/latest/getting_started.html>`_ instructions if you're new to the package manager.
|br|
|br|

> 
> 
> .. note::
> 
>   Before installing, you can ``spack spec``  variants to verify the build type.
> 
> Serial
> ~~~~~~~
> 
> .. code-block:: sh
> 
>   spack spec kokkos@4.2 %gcc@10.3.0 +serial cxxstd=20
> 
> OpenMP
> ~~~~~~
> 
> .. code-block:: sh
> 
>   spack spec kokkos@4.2 %gcc@10.3.0 +openmp cxxstd=20
> 
> 
> CUDA
> ~~~~
> 
> .. code-block:: sh
>   
>   spack spec / install kokkos@4.2 %gcc@12.2.0 +cuda cuda_arch=70 cxxstd=20 +cuda_relocatable_device_code
> 
> 
> HIP
> ~~~
> 
> .. code-block:: sh
> 
>   spack spec / install kokkos@4.2 %gcc@12.2.0 +rocm amdgpu_target=gfx90a cxxstd=20

Reviewer 5: ```suggestion
```

I am in favour to not give spack example and focus on Kokkos and a first example. Else we also have to describe how to find where Kokkos is installed and a lot of details.

Author: I am reluctant to remove the Spack recipes, as I don't think we have anything clearly showing a typical GPU-using variant with some degree of of complexity.  BUT, I take your point about linking to a Spack-installed Kokkos.  In the "Hello World" example, we could mention how one would link to the Spack-installed Kokkos.

Building and Linking a Kokkos "Hello World"
-------------------------------------------

.. note::

>   You will need to set ``Kokkos_ROOT``, and also the root directory for you target backend SDK (i.e., ``CUDA_ROOT``, ``ROCM_PATH``).  Please see `Build, Install and Use <https://kokkos.org/kokkos-core-wiki/building.html>`_ for additional details.

Reviewer 5: Not always for backends, they can be in various default paths or loaded via module. We do not talk about that when building Kokkos, why mention it now?
We can also use `CMAKE_PREFIX_PATH` or even `Kokkos_DIR` with CMake.

Author: I want to keep the CMake part of this as simple as possible, and dealing with things like `CMAKE_PREFIX_PATH` gets you into more complicated CMake territory (requiring and understanding of CMake modules).  I can fix the language to say these variables need to be set, no matter who sets them (i.e., system modules, or the user).

|br|

.. code-block:: sh

  git clone https://github.com/ajpowelsnl/View
  cd View

>   mkdir build
>   cd build
>   cmake ../

Reviewer 5: ```suggestion
  cmake -B build-dir -DKokkos_ROOT=<Kokkos_Install_Path> .
  cmake --build build-dir --parallel
```

Author: I want the "Quick Start" to focus on step-by-step, easy to follow Kokkos options, vs. those from CMake.   I want the focus to remain Kokkos, and not various CMake formulations, options, etc.

Reviewer 2: I agree in that at least we shouldn't have the `--parallel` there, to keep things minimal. `Kokkos_ROOT` is important as we need to specify it and it is more or less the "canonical way" in cmake to consume a package (`Kokkos_DIR` is more annoying to use, and `CMAKE_PREFIX_PATH` is really targeted towards full installations)

> ============
> Quick Start
> ============

Reviewer 2: ```suggestion
===========
Quick Start
===========
```

Author: Done!

>   This guide is intended to get new Kokkos users started in less than one hour.  Kokkos Core is the foundation for all other Kokkos Ecosystem projects.

Reviewer 2: ```suggestion
This guide is intended to get new Kokkos users started in less than one hour.  Kokkos Core is the foundation for all other Kokkos Ecosystem projects.
```

Identing is making it render weird (I think it believes its a quote)

------------
Requirements
------------

> ::

Reviewer 2: Change to a bullet list instead of code block

  CMake >=3.18
  gcc >= 8.3.1
  CUDA >= 11

>   rocm >=5.7

Author: This was my first guesstimate of requirements.  I'll update w/ the [requirements](https://kokkos.org/kokkos-core-wiki/requirements.html) in the documentation.

>   rocm >=5.7

Author: I linked to the existing requirements.

--------------------------------------
Download Current Release, Set Up Build 
--------------------------------------

..
  Choice 1 for obtaining current release:
..

> * `Click to Download Current Release <https://github.com/kokkos/kokkos/archive/refs/heads/master.zip>`_ 

Reviewer 2: This renders weirdly for me: 
<img width="821" alt="Screenshot 2024-02-19 at 5 01 55 PM" src="https://github.com/kokkos/kokkos-core-wiki/assets/6853627/a5b6ec80-3b0c-4d73-abbf-9ac506ded5f9">

> * `Click to Download Current Release <https://github.com/kokkos/kokkos/archive/refs/heads/master.zip>`_ 

Author: I believe I addressed both of your comments

..
  Temp. Notes --
  Buttons, badges, icons:
  https://sphinx-design.readthedocs.io/en/latest/badges_buttons.html
  :bdg-link-primary-line:`explicit title <https://example.com>`

  https://getbootstrap.com/docs/5.0/components/badge/

> ..

Author: Initially, yes 😄

..
  Choice 2 for obtaining current release:
..

> :bdg-link-primary:`Current Release <https://github.com/kokkos/kokkos/archive/refs/heads/master.zip>`

Author: I cleaned it up, and called the button "Latest" to agree with the github language.

> ::
> 
>   unzip kokkos-master.zip
>   cd kokkos-master
>   mkdir build
>   cd build

Reviewer 2: This isn't being syntax highlighted for me.

Also if you change the URL as I suggested you will need to change this:

```suggestion
.. code-block:: sh

  tar -xf kokkos-4.2.01.tar.gz
```
(Note the second URL I suggested gives a weird tarball name idk why)

>   cd build

Reviewer 4: Reiterating that I would not necessarily include the mkdir and cd

----------------------------------------------

> Basic Configuration Examples for Source Builds

Reviewer 2: ```suggestion
Basic Configuration
```

Author: Done!

----------------------------------------------

>   You can create small shell scripts to manage configuration details, following the GPU microarchitecture-appropriate examples below.  Upon successful configuration, ``make -j`` to build, and ``make install`` to install.

Reviewer 2: ```suggestion
You can create small shell scripts to manage configuration details, following the GPU microarchitecture-appropriate examples below.  Upon successful configuration, ``make -j`` to build, and ``make install`` to install.
```

Author: Done!

>   You can create small shell scripts to manage configuration details, following the GPU microarchitecture-appropriate examples below.  Upon successful configuration, ``make -j`` to build, and ``make install`` to install.

Reviewer 5: I really like cmake presets, do you think we can provide profiles for users or developers directly in Kokkos repository?

>   You can create small shell scripts to manage configuration details, following the GPU microarchitecture-appropriate examples below.  Upon successful configuration, ``make -j`` to build, and ``make install`` to install.

Author: With the "CMake presets" idea, I think we're getting into more complicated territory than I want w/ the Quick Start.  I do not want the Kokkos Quick Start to become a CMake tutorial.

Serial
~~~~~~

> ::

Reviewer 2: ```suggestion
.. code-block:: sh
```
And similar for any other shell code

  export INSTALL_DIR = ${MY_INSTALL_DIR}

  cmake \
    -DKokkos_ARCH_NATIVE=ON \
    -DKokkos_ENABLE_OPENMP=OFF \
    -DKokkos_ENABLE_SERIAL=ON \
    -DCMAKE_CXX_STANDARD=17 \
    -DCMAKE_INSTALL_PREFIX=$INSTALL_DIR \

>     -DCMAKE_BUILD_TYPE=Release ..

Author: I removed all `INSTALL_DIR` stuff, and got rid of `CXX_STANDARD`, as well.  I'm afraid I don't quite sure what to do with the comment about the `-DCMAKE_BUILD_TYPE` comment.  Can you be a bit more directive?

OpenMP
~~~~~~
::

  export INSTALL_DIR = ${MY_INSTALL_DIR}

  cmake \
   -DKokkos_ARCH_NATIVE=ON \
   -DKokkos_ENABLE_OPENMP=ON \
   -DKokkos_ENABLE_SERIAL=ON \
   -DCMAKE_CXX_STANDARD=17 \
   -DCMAKE_INSTALL_PREFIX=$INSTALL_DIR \
   -DCMAKE_BUILD_TYPE=Release ..

CUDA
~~~~

::

  export INSTALL_DIR = ${MY_INSTALL_DIR}

  cmake \
    -DKokkos_ARCH_NATIVE=ON \
    -DKokkos_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE=OFF \
    -DKokkos_ENABLE_CUDA=ON \
    -DKokkos_ENABLE_SERIAL=ON \
    -DCMAKE_CXX_STANDARD=17 \
    -DCMAKE_INSTALL_PREFIX=$INSTALL_DIR \
    -DCMAKE_BUILD_TYPE=Release ..

.. note::

  VEGA90A is for AMD MI200

>   AMD_GFX942 is for AMD MI300

Author: Sounds good

HIP
~~~
::

  cmake \

>     -DCRAYPE_LINK_TYPE=dynamic \

Reviewer 4: That does not seem like the canonical HIP configuration.  If you want to give configuration for some LCF system then say what system it is for.

    -DKokkos_ENABLE_TESTS=ON \
    -DKokkos_ENABLE_HIP=ON \
    -DKokkos_ARCH_AMD_GFX942=OFF \
    -DKokkos_ARCH_VEGA90A=ON \
    -DKokkos_ENABLE_SERIAL=ON \
    -DCMAKE_CXX_STANDARD=17 \

>     -DCMAKE_CXX_COMPILER=$(which hipcc) \

Reviewer 4: Why `$(which hipcc)` rather than just `hipcc`?

    -DCMAKE_INSTALL_PREFIX=$INSTALL_DIR \
    -DCMAKE_BUILD_TYPE=Release ..

----------------------------------------------
Basic Configuration Examples for Spack Builds
----------------------------------------------

*  A number of Kokkos variants can be built / installed via Spack.  You will need to select a variant for the desired backend, and appropriate GPU microarchitecture.   
*  To explore the range of variants for a package, ``spack info kokkos``, ``spack info trilinos``, etc.

> * `Basic Spack Installation <https://spack.readthedocs.io/en/latest/getting_started.html>`_

Reviewer 2: I think this would read better in prose rather than a bulleted list

Author: Done

.. note::

  Before installing, you can ``spack spec``  variants to verify the build type.

Serial
~~~~~~~

::

>   spack spec kokkos@4.2 %gcc@10.3.0 +serial cxxstd=20

Reviewer 2: Please fix this

OpenMP
~~~~~~

::

  spack spec kokkos@4.2 %gcc@10.3.0 +openmp cxxstd=20

CUDA
~~~~

:: 
  
  spack spec / install kokkos@4.2 %gcc@12.2.0 +cuda cuda_arch=70 cxxstd=20 +cuda_relocatable_device_code

HIP
~~~

::

>   spack spec / install kokkos@4.2 %gcc@12.2.0 +rocm amdgpu_target=gfx90a cxxstd=20

Reviewer 4: Do we really want yet another Spack section?  Can';t we just refer to the existing one?

Author: RE - refer to existing; the existing does not give examples of the more complicated build formulations.  Also, the point of the "Quick Start" is to get going quickly.  The existing "build,install and use" guide is a lot of words.

--------------------
Additional Resources
--------------------

* `CMake Keywords <https://kokkos.org/kokkos-core-wiki/keywords.html>`_
* `Building Kokkos <https://kokkos.org/kokkos-core-wiki/building.html>`_
* `Spack Kokkos Build <https://kokkos.org/kokkos-core-wiki/building.html#spack>`_

> 

Reviewer 2: Link to our tutorials here?

Author: Done!

---
FAQ
---

* How to Join Kokkos Slack?

You can find the Slack channel at `kokkosteam.slack.com <https://kokkosteam.slack.com>`_. Register a new account with your email. We automatically whitelist emails from most organizations, but if your email address is not whitelisted, you can contact the Kokkos maintainers (their emails are in the LICENSE file).

* How to build for more recent C++??

When configuring Kokkos with CMake (see configruation, add the flag ``-DCMAKE_CXX_STANDARD=20`` (or ``23``). Ensure that the flag is also set for any downstream applications.

> 

Reviewer 4: Why are we repeating the content of the FAQ?
Did you just copy the file and forgot to remove?

Author: I removed it.  A little redundancy is not bad, but in subsequent edits, I agree with you -- it just clutters.

> ============

Reviewer 2: Please use the heading scheme in the comments at the top of https://raw.githubusercontent.com/kokkos/kokkos-core-wiki/main/docs/source/templates/class_api.rst

(I realized in hindsight it's not the most accessible as it is in a comment and not rendered)

Author: Done!

Quick Start
============

This guide is intended to jump start new Kokkos users.  Kokkos Core is the foundation for all other Kokkos Ecosystem projects, and is readily usable in your own for performance portability.

-----------------------------
Download Latest, Set Up Build 
-----------------------------

.. note::

  Please become familiar with `Kokkos Requirements <https://kokkos.org/kokkos-core-wiki/requirements.html>`_, and verify that your machine has all necessary compilers, backend SDK and build system components.

..
 Nota bene:  the link for "Latest" will need to be updated for each release 
..

> :bdg-link-primary:`Latest Release <https://github.com/kokkos/kokkos/releases/tag/4.2.01>`

Author: Done!

> ::

Reviewer 2: One more thing I forgot is that you should use a syntax highlighting language here:

e.g.
```rst
.. code-block:: sh
```

See https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-code-block

Author: Done!

  unzip kokkos-x.y.z.zip (or tar -xzf kokkos-x.y.z.tar.gz)
  cd kokkos-x.y.z
  mkdir build
  cd build

-------------------
Basic Configuration
-------------------

> .. note::
> 
>   You can create small shell scripts to manage and experiment with configuration details, following the GPU microarchitecture-appropriate examples below.  Upon successful configuration, ``make -j`` to build, and ``make install`` to install.

Reviewer 2: ```suggestion
You can create small shell scripts to manage and experiment with configuration details, following the GPU microarchitecture-appropriate examples below.  Upon successful configuration, ``make -j`` to build, and ``make install`` to install.
```

Author: For the `note` inset to work, you must indent.

Reviewer 2: I think here it makes more sense to have it prose rather than a note

Serial (CPU)
~~~~~~~~~~~~
::

>   cmake \
>     -DKokkos_ARCH_NATIVE=ON \
>     -DKokkos_ENABLE_OPENMP=OFF \
>     -DKokkos_ENABLE_SERIAL=ON \
>     -DCMAKE_BUILD_TYPE=Release ..

Reviewer 2: ```suggestion
  cmake -S . -B build -DCMAKE_BUILD_TYPE=Release
```

Where build type is optional...

But I agree with Damien, I don't think there is much point in showing a serial build here

Author: OK, I removed `Serial` ... for now ... but I'm not entirely sure I agree with you.  The point of the "Quick Start" is to cast a wide net for a range of experience levels, and not just seasoned parallel programmers.  Also, the point of showing the "Quick Start" in this manner is to give newer people an idea about the recurrent themes you see working with Kokkos.

Reviewer 2: I think it's worth bringing up in the documentation meeting or the team meeting if it's more urgent

.. note::

  Kokkos will attempt to autodetect GPU microarchitecture, but it is also possible to specify the desired `GPU architecture <https://kokkos.org/kokkos-core-wiki/keywords.html#gpu-architectures>`_.   

OpenMP (CPU Parallelism)
~~~~~~~~~~~~~~~~~~~~~~~~
::

>   cmake \
>    -DKokkos_ARCH_NATIVE=ON \
>    -DKokkos_ENABLE_OPENMP=ON \
>    -DKokkos_ENABLE_SERIAL=ON \
>    -DCMAKE_BUILD_TYPE=Release ..

Reviewer 2: ```suggestion
  cmake -S . -B build -DKokkos_ENABLE_OPENMP=ON -DCMAKE_BUILD_TYPE=Release
```

Author: I prefer the current version.  Plus, the current version is continuous with earlier parts of the document.

> CUDA (CPU and GPU)
> ~~~~~~~~~~~~~~~~~~

Reviewer 2: ```suggestion
CUDA
~~~~
```

::

>   cmake \
>     -DKokkos_ARCH_NATIVE=ON \
>     -DKokkos_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE=OFF \
>     -DKokkos_ENABLE_CUDA=ON \
>     -DKokkos_ENABLE_SERIAL=ON \
>     -DCMAKE_BUILD_TYPE=Release ..

Reviewer 2: ```suggestion
  cmake -S . -B -DKokkos_ENABLE_CUDA=ON -DCMAKE_BUILD_TYPE=Release
```

BTW if nvcc is not in their path this will fail (but I think most clusters will load cuda as an environment module)

Author: I prefer the current version.

Reviewer 2: I missed a `-B build` in there but regardless, if you prefer the `cmake ..` approach (I'm not hard set either way), I still think it's important to remove the ARCH_NATIVE, the CUDA_RELOCATABLE_DEVICE_CODE and the ENABLE_SERIAL flags as they are not required for building.

> HIP (CPU and GPU)
> ~~~~~~~~~~~~~~~~~

Reviewer 2: ```suggestion
HIP
~~~
```

::

  cmake \
    -DKokkos_ARCH_NATIVE=ON \
    -DKokkos_ENABLE_HIP=ON \
    -DKokkos_ENABLE_SERIAL=ON \
    -DCMAKE_CXX_COMPILER=hipcc \

>     -DCMAKE_BUILD_TYPE=Release ..

Reviewer 2: Same as above but not 100% sure if hipcc needs to be specified directly here

----------------------------------------------
Basic Configuration Examples for Spack Builds
----------------------------------------------

> |br|
> Diverse Kokkos variants can be built / installed via Spack.  You will need to select a variant for the desired backend, and appropriate GPU microarchitecture.  To explore the range of variants for a package, ``spack info kokkos``, ``spack info trilinos``, etc.  Please see basic `Spack installation  <https://spack.readthedocs.io/en/latest/getting_started.html>`_ instructions if you're new to the package manager.
> |br|
> |br|

Reviewer 2: ```suggestion
Diverse Kokkos variants can be built / installed via Spack.  You will need to select a variant for the desired backend, and appropriate GPU microarchitecture.  To explore the range of variants for a package, ``spack info kokkos``, ``spack info trilinos``, etc.  Please see basic `Spack installation  <https://spack.readthedocs.io/en/latest/getting_started.html>`_ instructions if you're new to the package manager.
```

Remove extraneous line breaks

.. note::

  Before installing, you can ``spack spec``  variants to verify the build type.

Serial
~~~~~~~

::

  spack spec kokkos@4.2 %gcc@10.3.0 +serial cxxstd=20

OpenMP
~~~~~~

::

  spack spec kokkos@4.2 %gcc@10.3.0 +openmp cxxstd=20

CUDA
~~~~

:: 
  
  spack spec / install kokkos@4.2 %gcc@12.2.0 +cuda cuda_arch=70 cxxstd=20 +cuda_relocatable_device_code

HIP
~~~

::

  spack spec / install kokkos@4.2 %gcc@12.2.0 +rocm amdgpu_target=gfx90a cxxstd=20

-------------------------------------------
Building and Linking a Kokkos "Hello World"
-------------------------------------------

.. note::

  You will need to set ``Kokkos_ROOT``, and also the root directory for you target backend SDK (i.e., ``CUDA_ROOT``, ``ROCM_PATH``).  Please see `Build, Install and Use <https://kokkos.org/kokkos-core-wiki/building.html>`_ for additional details.

> |br|

Reviewer 2: ```suggestion
```

::

>   git clone https://github.com/ajpowelsnl/View

Reviewer 2: I think this template project should be in the Kokkos org

  cd View
  mkdir build
  cd build
  cmake ../

-------------
Get Involved!
-------------

Joining the Kokkos Community
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Participate in the lively virtual discussion on the `Kokkos Slack Channel <https://kokkosteam.slack.com>`_.  Here are `sign up details <https://kokkos.org/kokkos-core-wiki/faq.html#faq>`_.

> |br|

Reviewer 2: ```suggestion
```

Acclerating learning
~~~~~~~~~~~~~~~~~~~~

Take a deeper dive into Kokkos with over 16 hours of `Tutorials <https://github.com/kokkos/kokkos-tutorials>`_ and `Recorded Lectures <https://github.com/kokkos/kokkos-tutorials/wiki/Kokkos-Lecture-Series>`_.  For in-house workshops and training, please get in touch via Slack (below).

> |br|

Reviewer 2: ```suggestion
```

Coming to Release Demos
~~~~~~~~~~~~~~~~~~~~~~~

You have a standing invitation to attend our approximately quarterly release demos to hear what's new, what's gone, and where we're heading in the future.

> |br|

Reviewer 2: Don't we call these release briefings?

> Coming to Release Demos
> ~~~~~~~~~~~~~~~~~~~~~~~
> 
> You have a standing invitation to attend our approximately quarterly release demos to hear what's new, what's gone, and where we're heading in the future.
> |br|

Reviewer 2: ```suggestion
Coming to Release Demos
~~~~~~~~~~~~~~~~~~~~~~~

You have a standing invitation to attend our approximately quarterly release demos to hear what's new, what's gone, and where we're heading in the future.
```

Author: I introduced the `|br|` for readability, and would like to leave them there.

Attending Kokkos Users' Group Meetings
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Please join us in our annual Kokkos Users' Group Meeting, where we present recent Kokkos work, and you showcase how you use Kokkos.  It's a great opportunity to build community and grow collaboration.

> |br|

Reviewer 2: ```suggestion
```

..
  *TODO*
     - Integrate (merged) Quick Start with Cédric's PR:  https://github.com/kokkos/kokkos/pull/6796
     - Ongoing reconciling with the Julien B. / KUG23- initiated discussion:  https://github.com/kokkos/internal-documents/pull/19
     - Add `git submodule` "how to" for Kokkos
     - Add Quick Start to main Kokkos page, such that it is the first thing you encounter on the landing page (kokkos.org)
     - In V2, put the recipes for the different backends on different pages
     - Julien B. suggested using github templates for the View "Hello World" example
     - Nic M.:  CUDA as a CMake language example (using View): cmake -S . -B build -DKokkos_ENABLE_CUDA=ON CMAKE_CUDA_COMPILER=nvcc Kokkos_ENABLE_COMPILE_AS_CMAKE_LANGUAGE=ON [-DCMAKE_BUILD_TYPE=Release]

> ..

Reviewer 2: ```suggestion
```

Author: The current version does not have 194 lines 😄 

Those two dots are part of a comment that I would like to leave in V1 until the Quick Start is integrated with items in the list.  The comment has a half life (relatively short).

> .. |br| raw:: html
> 
>       <br>
> 

Reviewer 2: ```suggestion
```

Author: This html-ism is needed for new lines (`|br|`) in Sphinx.  New lines are needed in certain places because the text is crowded with too many words (towards the bottom).  I would like to leave this html-ism, unless there's a better way to do this in Sphinx.
